codalab/lib/print_util.py

```diff
@@ -12,3 +15,45 @@ def clear_line(f=sys.stderr):
 
 def pretty_print(obj, f=sys.stdout):
     json.dump(obj, f, sort_keys=True, indent=4, separators=(',', ': '))
+
+
+class FileTransferProgress(object):
+    """
+    Formats and displays progress on operations involving transferring bytes.
+
+    Should be used as a context manager:
+
+        with FileTransferProgress() as progress:
+            while 1:
+                ...
+                progress.update('Uploading ', num_bytes, total_bytes)
+                ...
+    """
+    def __init__(self, f=sys.stderr):
+        self.f = f
+
+    def __enter__(self):
+        self.start_time = time.time()
+        return self
+
+    def update(self, msg, bytes_done, bytes_total=None):
+        """
+        Update progress display.
+
+        :param msg: Message to prepend the progress text.
+        :param bytes_done: Number of bytes transferred
+        :param bytes_total: Number of bytes total to transfer, or None if unknown
+        """
+        self.f.write('\r')
+        self.f.write(msg)
+        if bytes_total is None:
+            self.f.write(formatting.size_str(bytes_done))
+        else:
+            self.f.write(formatting.ratio_str(formatting.size_str, bytes_done, bytes_total))
+        speed = float(bytes_done) / (time.time() - self.start_time)
+        self.f.write(' [%sbps]' % formatting.size_str(speed))
```

Comment:
This prints out 'mbps' which is not what we want - I'd go with MB/sec explicitly (don't need formatting.size_str, since the dynamic range here isn't going to be that big)